### PR TITLE
Added notification to all components when changes will be lost

### PR DESCRIPTION
**Pull request by dadamek**
Task: Provide warning to user when changes will be lost (8443b430-6d7a-11ec-8419-fbd10f9c9d1c)

### CODE REVIEW

**Comment by tpitts-CFI** [comment]
> * In Admin component, Journal Component => If I select to discard the changes, the Dirty Indicator remains until I navigate away from the selected item
> * Modifying an item on the Project Status -> Board (in dashboard) does not provided a warning before discard.

This is not an issue with this task. It is being fixed with two other tasks:
* Task: Handle creates, deletes and updates in session service (66617cd0-6f29-11ec-8419-fbd10f9c9d1c)
* Task: User and Session Push Notification (aa7df250-e001-11e8-a3f0-ed214af03fd6)

File: kohese-portal/client/src/components/document-view/document-row/document-row.component.ts [hunk 1]
```diff
@@ -184,10 +184,23 @@ export class DocumentRowComponent implements OnInit, OnDestroy, AfterViewInit {
...
+        this.row.editable = true;
+      }
+      if(response === true) {
+        await this._itemRepository.fetchItem(this.docInfo.proxy);
+        this.row.editable = false;
+        this.checkEntries();
+        this.changeRef.markForCheck();
+      }
+    } else {
+      await this._itemRepository.fetchItem(this.docInfo.proxy);
```
Comment (tpitts-CFI) [inline kohese-portal/client/src/components/document-view/document-row/document-row.component.ts:199]: line 199 may be removed

File: kohese-portal/client/src/components/dashboard/assignment-dashboard/assignment-dashboard.component.ts [hunk 2]
```diff
@@ -35,6 +35,7 @@ import { TreeConfiguration } from '../../../../../common/src/tree-configuration'
 import { DashboardSelections } from '../dashboard-selector/dashboard-selector.component';
 import { FormatObjectEditorComponent } from '../../object-editor/format-object-editor/format-object-editor.component';
 import { Hotkeys } from '../../../services/hotkeys/hot-key.service';
+import { proxy } from 'jquery';
```
Comment (tpitts-CFI) [inline kohese-portal/client/src/components/dashboard/assignment-dashboard/assignment-dashboard.component.ts:38]: remove this import

**Review by tpitts-CFI — CHANGES_REQUESTED** [review]
Tpitts review

File: kohese-portal/client/src/components/dashboard/project-dashboard/status-dashboard/state-bar-chart/state-summary-dialog/state-summary-dialog.component.ts [hunk 3]
```diff
@@ -184,12 +184,21 @@ export class StateSummaryDialogComponent implements OnInit, OnDestroy {
...
-    this.checkEntries(itemProxy);
-    this.changeRef.markForCheck();
+  public async discardChanges(itemProxy: ItemProxy): Promise<void> {
+    if(itemProxy.dirty) {
+      let response = await this._dialogService.openYesNoDialog('Discard Changes?', '');
+      if(response === false) {
+        return;
+      }
+      if(response === true) {
+        await this._itemRepository.fetchItem(TreeConfiguration.getWorkingTree().getProxyFor(itemProxy.item.id));
```
[the remark this answers is not in the document]
Comment (dadamek) [reply kohese-portal/client/src/components/dashboard/project-dashboard/status-dashboard/state-bar-chart/state-summary-dialog/state-summary-dialog.component.ts:194]: fixed

File: kohese-portal/client/src/components/dashboard/assignment-dashboard/assignment-dashboard.component.ts [hunk 4]
```diff
@@ -182,11 +183,20 @@ export class AssignmentDashboardComponent implements OnInit, OnDestroy {
...
-    this._editableSet.splice(this._editableSet.indexOf(itemProxy.item.id), 1);
-    this.checkEntries(itemProxy);
-    this.changeRef.markForCheck();
+    if(itemProxy.dirty) {
+      let response = await this._dialogService.openYesNoDialog('Discard Changes?', '');
+      if(response === false) {
+        return;
+      }
+      if(response === true) {
+        await this._itemRepository.fetchItem(TreeConfiguration.getWorkingTree().getProxyFor(itemProxy.item.id));
```
[the remark this answers is not in the document]
Comment (dadamek) [reply kohese-portal/client/src/components/dashboard/assignment-dashboard/assignment-dashboard.component.ts:192]: fixed

File: kohese-portal/client/src/components/details/journal/journal.component.ts [hunk 5]
```diff
@@ -268,13 +268,43 @@ export class JournalComponent implements OnDestroy {
...
-    this._editableSet.splice(this._editableSet.indexOf(itemProxy.item.id), 1);
-    this._changeDetectorRef.markForCheck();
+  public async discardChanges(itemProxy: ItemProxy): Promise<void> {
+    if(itemProxy.dirty) {
+      let response = await this._dialogService.openYesNoDialog('Discard Changes?', '');
+      if(response === false) {
+        return;
+      }
+      if (response === true) {
+        this._itemRepository.fetchItem(TreeConfiguration.getWorkingTree().getProxyFor(itemProxy.item.id));
```
[the remark this answers is not in the document]
Comment (dadamek) [reply kohese-portal/client/src/components/details/journal/journal.component.ts:278]: fixed

File: kohese-portal/client/src/components/details/journal/journal.component.ts [hunk 6]
```diff
@@ -268,13 +268,43 @@ export class JournalComponent implements OnDestroy {
...
+        this._editableSet.splice(this._editableSet.indexOf(itemProxy.item.id), 1);
+        this._changeDetectorRef.markForCheck();
+      }
+    } else {
+      this._editableSet.splice(this._editableSet.indexOf(itemProxy.item.id), 1);
+    }
   }
 
+
```
[the remark this answers is not in the document]
Comment (dadamek) [reply kohese-portal/client/src/components/details/journal/journal.component.ts:287]: fixed

File: kohese-portal/client/src/components/document-view/document-row/document-row.component.ts (hunk 1 above)
Comment (dadamek) [reply kohese-portal/client/src/components/document-view/document-row/document-row.component.ts:199]: fixed

File: kohese-portal/client/src/components/dashboard/assignment-dashboard/assignment-dashboard.component.ts (hunk 2 above)
Comment (dadamek) [reply kohese-portal/client/src/components/dashboard/assignment-dashboard/assignment-dashboard.component.ts:38]: fixed

**Review by tpitts-CFI — APPROVED** [review]
all comments were addressed. I am okay with this merge request.

**Review by gmckune — CHANGES_REQUESTED** [review]
1.  Project Status board is working.  Changes were discarded and Dirty flag went away.
2. Journal Component still not working.  Dirty flag doesn't go away until mouse is moved and changes remain that I wanted discarded.  But, the changes that I wanted discarded disappear if I click edit again.
3. Admin changes were not retested.  However, need to determine which task will fix the problem here (since two tasks are identified).

**Review by gmckune — APPROVED** [review]

**Pull request merged**